### PR TITLE
add p2p functions for sync popmempool after nodes sync

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: src/vbk/p2p_sync.hpp [hunk 1]
```diff
@@ -7,27 +7,127 @@
...
+    AssertLockHeld(cs_main);
+    auto& pop_mempool = VeriBlock::getService<VeriBlock::PopService>().getMemPool();
+    auto data = pop_mempool.getMap<PopDataType>();
+
+    auto& known_set = mapPopDataNodeState[node->GetId()].getSet<PopDataType>();
+
+    std::vector<std::vector<uint8_t>>
+        hashes;
+    for (const auto& el : data) {
+        if (known_set.find(el.first) == known_set.end()) {
```
Comment (Warchant) [inline src/vbk/p2p_sync.hpp]:
```suggestion
        if (known_set.count(el.first) == 0) {
```

File: src/vbk/p2p_sync.cpp [hunk 2]
```diff
@@ -1,57 +1,145 @@
...
 #include "veriblock/entities/vbkblock.hpp"
 #include "veriblock/entities/vtb.hpp"
 
-#include "veriblock/mempool.hpp"
-
 namespace VeriBlock {
 namespace p2p {
 
-bool processPopData(CNode* pfrom, const std::string& strCommand, CDataStream& vRecv, int64_t nTimeReceived, const CChainParams& chainparams, CConnman* connman, BanMan* banman, const std::atomic<bool>& interruptMsgProc)
+std::map<NodeId, PopDataNodeState> mapPopDataNodeState;
```
Comment (Warchant) [inline src/vbk/p2p_sync.cpp]:
```suggestion
std::map<NodeId, std::shared_ptr<PopDataNodeState>> mapPopDataNodeState;
```

File: src/vbk/p2p_sync.cpp [hunk 3]
```diff
@@ -1,57 +1,145 @@
...
+
+        if (requested_data.size() == MAX_POP_DATA_SENDING_AMOUNT) {
+            connman->PushMessage(node, msgMaker.Make(get_prefix + PopDataType::name(), requested_data));
+            requested_data.clear();
+        }
+    }
+
+    if (!requested_data.empty()) {
+        connman->PushMessage(node, msgMaker.Make(get_prefix + PopDataType::name(), requested_data));
+    }
```
Comment (Warchant) [inline src/vbk/p2p_sync.cpp:64]: you can do PushMessage once, as requested_data in the loop will never be > MAX_POP_DATA_SENDING_AMOUNT

File: src/vbk/p2p_sync.cpp [hunk 4]
```diff
@@ -1,57 +1,145 @@
...
+bool processPopData(CNode* node, CDataStream& vRecv, altintegration::MemPool& pop_mempool) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+{
+    AssertLockHeld(cs_main);
+    LogPrint(BCLog::NET, "received pop data: %s, bytes size: %d", PopDataType::name(), vRecv.size());
+    PopDataType data;
+    vRecv >> data;
+
+    altintegration::ValidationState state;
+    if (!pop_mempool.submit(data, state)) {
+        LogPrint(BCLog::NET, "VeriBlock-PoP: %s ", state.GetPath());
```
Comment (Warchant) [inline src/vbk/p2p_sync.cpp]: add meaningful message
```suggestion
        LogPrint(BCLog::NET, "peer %d sent statelessly invalid pop data: %s ", <PEERNUMBER>, state.GetPath());
```

File: src/vbk/p2p_sync.hpp [hunk 5]
```diff
@@ -7,27 +7,129 @@
...
+const static std::string offer_prefix = "of";
+
+const static uint32_t MAX_POP_DATA_SENDING_AMOUNT = MAX_INV_SZ;
+
+template <typename PopDataType>
+void offerPopData(CNode* node, CConnman* connman, const CNetMsgMaker& msgMaker) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+{
+    AssertLockHeld(cs_main);
+    auto& pop_mempool = VeriBlock::getService<VeriBlock::PopService>().getMemPool();
+    auto data = pop_mempool.getMap<PopDataType>();
```
Comment (Warchant) [inline src/vbk/p2p_sync.hpp]:
```suggestion
    const auto& data = pop_mempool.getMap<PopDataType>();
```

File: src/vbk/p2p_sync.hpp [hunk 6]
```diff
@@ -7,27 +7,129 @@
...
+        hashes;
+    for (const auto& el : data) {
+        if (known_set.find(el.first) == known_set.end()) {
+            hashes.push_back(el.first.asVector());
+            known_set.insert(el.first);
+        }
+
+        if (hashes.size() == MAX_POP_DATA_SENDING_AMOUNT) {
+            connman->PushMessage(node, msgMaker.Make(offer_prefix + PopDataType::name(), hashes));
+            data.clear();
```
Comment (Warchant) [inline src/vbk/p2p_sync.hpp]:
```suggestion
            hashes.clear();
```

File: src/vbk/p2p_sync.hpp (hunk 1 above)
Comment (Mr-Leshiy) [reply src/vbk/p2p_sync.hpp]: fixed

File: src/vbk/p2p_sync.cpp (hunk 3 above)
Comment (Mr-Leshiy) [reply src/vbk/p2p_sync.cpp:64]: fixed

File: src/vbk/p2p_sync.hpp (hunk 5 above)
Comment (Mr-Leshiy) [reply src/vbk/p2p_sync.hpp]: fixed

File: src/vbk/p2p_sync.hpp (hunk 6 above)
Comment (Mr-Leshiy) [reply src/vbk/p2p_sync.hpp]: fixed

File: src/vbk/p2p_sync.cpp (hunk 4 above)
Comment (Mr-Leshiy) [reply src/vbk/p2p_sync.cpp]: fixed

File: src/vbk/p2p_sync.cpp (hunk 2 above)
Comment (Mr-Leshiy) [reply src/vbk/p2p_sync.cpp]: fixed

**Review by Warchant — APPROVED** [review]

**Pull request merged**